src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,296 @@
+                .AddReferences(await NuGetReferenceProvider.GetReferences())
+                .WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            _semanticModel = _compilation.GetSemanticModel(_syntaxTree, false);
+            var semanticDiagnostics = _semanticModel.GetDiagnostics();
+
+            validationMessages.AddRange(semanticDiagnostics.Select(DiagnosticToValidationMessage));
+
+            return CreateAndSetValidationResult(out _syntaxValidationResult,
+                !ErrorExists(validationMessages),
+                validationMessages);
+        }
+
+        public Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var algoClassFullName = _syntaxWalker.NamespaceNode == null
+                ? $"{_syntaxWalker.ClassNode.Identifier}"
+                : $"{_syntaxWalker.NamespaceNode.Name}.{_syntaxWalker.ClassNode.Identifier}";
+
+            var newAssembly = GenerateAssembly();
+            var newType = newAssembly.GetType(algoClassFullName);
+            var metadata = ExtractMetadata(newType);
+
+            return Task.FromResult(metadata);
+        }
+
+        private AlgoMetaDataInformation ExtractMetadata(Type algoType)
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            //Get public properties that have public setter
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                .Where(x => x.GetSetMethod() != null);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                //Base parameters
+                if (algoProperty.PropertyType.BaseType != typeof(AbstractFunction)
+                    && !typeof(IFunction).IsAssignableFrom(algoProperty.PropertyType))
+                {
+                    var parameter = ToAlgoMetaDataParameter(algoProperty);
+
+                    if (algoProperty.PropertyType.IsEnum)
+                        parameter.PredefinedValues = ToEnumValues(algoProperty);
+
+                    metadata.Parameters.Add(parameter);
+                }
+                //Functions
+                else if (algoProperty.PropertyType.BaseType == typeof(AbstractFunction)
+                         || typeof(IFunction).IsAssignableFrom(algoProperty.PropertyType))
+                {
+                    //Get first public property that is not function base parameters but it inherits it
+                    var functionProperty = algoProperty.PropertyType
+                        .GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly |
+                                       BindingFlags.GetProperty |
+                                       BindingFlags.SetProperty) //BindingFlags.DeclaredOnly -> ignore inherited members
```

Comment:
`BindingFlags.GetProperty` and `BindingFlags.SetProperty` are only used in `InvokeMember` calls, so they are unneeded here (as they won't have any effect). Please remove them from anywhere they appear outside of `InvokeMember`